### ISSUE TITLE
Lint against `stream.next().fuse()`

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust-clippy#11143]
### What it does

If [`FutureExt::fuse`] is called on a [`stream::Next`], suggest removing the call. If the underlying stream does not implement [`FusedStream`], additionally suggest calling [`StreamExt::fuse`] on the stream once it is created (or at the beginning of a function, if it was passed as an argument).

[`FusedStream`]: https://docs.rs/futures/latest/futures/stream/trait.FusedStream.html
[`FusedFuture`]: https://docs.rs/futures/latest/futures/future/trait.FusedFuture.html
[`FutureExt::fuse`]: https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse
[`StreamExt::fuse`]: https://docs.rs/futures/0.3.28/futures/stream/trait.StreamExt.html#method.fuse
[`stream::Next`]: https://docs.rs/futures/latest/futures/stream/struct.Next.html

### Advantage

The `select!` macro requires every future it polls to implement [`FusedFuture`], which adds an `is_terminated` method that tracks whether the future should continue to be polled.

Sometimes to satisfy this requirement, a person will fuse the `Next` future for a stream rather than fusing the stream itself. This is likely to be wrong, because `Next` is only an ephemeral future that waits for a single item from the stream and carries no state of its own.

Fusing a future adds memory of whether that future has returned an item or not, but _not_ whether the stream itself has terminated. If the select is happening in a loop, as is often the case, this can lead to panics because streams are allowed to panic if they are polled after they terminate.

`Next` already implements `FusedFuture` if the stream itself implements [`FusedStream`](https://docs.rs/futures/latest/futures/stream/trait.FusedStream.html). If it does not, the stream can be fused using the `StreamExt::fuse` adapter. This pushes the termination tracking up to the stream itself rather than into one ephemeral future, which is almost certainly what the user wants.

[`FusedFuture`]: https://docs.rs/futures/latest/futures/future/trait.FusedFuture.html
[`FutureExt::fuse`]: https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse
[`StreamExt::fuse`]: https://docs.rs/futures/0.3.28/futures/stream/trait.StreamExt.html#method.fuse
[`stream::Next`]: https://docs.rs/futures/latest/futures/stream/struct.Next.html

### Not harmless

As stated above, this pattern can lead to unexpected panics. Sometimes it would be considered harmless, if suboptimal, if an already-fused future is fused again to be used in a select. That is not the case here.

The reason is that a stream, even one that implements `FusedFuture`, is allowed to panic if it is polled after termination. This is very subtle because that the `Fuse` adapters (for futures and streams respectively) protect against this by refusing to poll their inner objects, but this behavior is _not_ actually required by the traits themselves.

Unnecessarily fusing a `Next` future of a stream that implements `FusedStream` "overrides" its existing `FusedFuture` implementation and causes it to ignore the termination status of the stream itself. This will lead to the future, and therefore the underlying stream, getting polled unconditionally by the select, which (if the stream is already terminated) can lead to a panic.

Of course, if a stream is already known not to panic if it is polled after termination, this pattern can still be harmless. But it is still brittle and suboptimal.

### Drawbacks

Maybe there are situations where this is actually desired, but I think it's unlikely. In every case you should be able to fuse the stream, or take a mutable reference to the stream and fuse that instead.

### Example

```rust
async fn test_watch() {
    let env = TestEnv::new();
    let mut watch_stream = env.repo.watch().unwrap();

    futures::select! {
        _ = watch_stream.next().fuse() /* <- bad! */ => panic!("should not have received an event"),
        _ = Timer::new(Duration::from_millis(10)).fuse() => (),
    };
}
```

Could be written as:

```rust
async fn test_watch() {
    let env = TestEnv::new();
    let mut watch_stream = env.repo.watch().unwrap().fuse(); // <- good!

    futures::select! {
        _ = watch_stream.next() => panic!("should not have received an event"),
        _ = Timer::new(Duration::from_millis(10)).fuse() => (),
    };
}
```
